Ch. Part. API: etcdraft consenter w/o system channel

tock-ibm opened the pull request:
#### Type of change

- New feature

#### Description

This PR contains 3 commits in order to make it easier to follow. 

Part 1 of 3 commits: Refactor onboarding out of server.
Here the onboarding code is refactored out of the common/server package in to an (almost) leaf package that depends mainly on cluster. This is done in order to make server     main.go cleaner and to allow the multichannel as well as other packages to reuse the onboarding code (break the import cycle). This commit introduces no functional change.

    
Part 2 of 3 commits: Refactor main to setup cluster when no system channel.
Here the server main.go is refactored in order to make it clearer and simpler, and setup cluster when there is no system channel and no bootstrap block. We assume that is this case the network is of cluster type, i.e. etcdraft, and therefore setup everything we can to allow a correct initialization of the cluster.

Part 3 of 3 commits: Create an etcdraft consenter when there is no system channel.
Here the intialization of an etcdraft consenter in main.go is performed even when there is no system channel and no bootstrap block. The orderer can now start without channels at  all and have an initialized consenter, (almost) ready for the creation of new channels.

#### Related issues

Task: FAB-17961
Story: FAB-15711
Epic: FAB-17712

srderson commented:
LGTM. Comments and whitespace only, but also tests have run and passed.

tock-ibm committed:
FAB-17961 Ch.Part.API: Refactor onboarding out of server

Part 1 of 3 commits: here the onboarding code is refactored out
of the common/server package in to an (almost) leaf package that
depends mainly on cluster. This is done in order to make it server
main() cleaner and to allow the multichannel as well as other
packages to reuse the onboarding code (break the import cycle).

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: Ia8bea790132d3ee6e182a0a5e0dc9b0da40c731f

tock-ibm committed:
FAB-17961 Ch.Part.API: Refactor main to setup cluster when no system channel

Part 2 of 3 commits: here the server main.go is refactored in order to make
it clearer and simpler, and setup cluster when there is no system channel
and no bootstrap block. We assume that is this case the network is of
cluster type, i.e. etcdraft, and therefore setup everything we can to
allow a correct initialization of the cluster.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I26352d1e8a04082fcc7d1e96b8169990d7818c05

tock-ibm committed:
FAB-17961 Ch.Part.API: Create an etcdraft consenter when there is no system channel

Part 3 of 3 commits: here the intialization of an etcdraft consenter
in main.go is performed even when there is no system channel and
no bootstrap block. The orderer can now start without channels at
all and have an initialized consenter,  ready for the creation of
new channels.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I20caf39e33395447b8d1a5c8f606948bed2ccb94

wlahti commented on orderer/common/server/main.go:
```diff
@@ -112,58 +112,91 @@ func Main() {
 			}
 		}
+	} else {
+		logger.Infof("Setting up cluster when there is no system channel: assuming cluster-type consensus")
```
"assuming cluster-type consensus" seems like an odd thing to include in an info message, especially when the first part of the message mentions we're setting up a cluster.

tock-ibm replied:
Yes, it sounds weird. Any suggestions?

wlahti replied:
Just seems redundant to include that last part is what I was getting at. I'd drop everything after system channel. And also reword "when there is no system channel" to "without a system channel" to be consistent with the other existing log messages.

tock-ibm replied:
ok

wlahti commented on orderer/common/server/main.go:
```diff
@@ -331,6 +364,11 @@ func selectClusterBootBlock(bootstrapBlock, sysChanLastConfig *cb.Block) *cb.Blo
 	}
 
+	if bootstrapBlock == nil {
+		logger.Info("Selected system channel last config block, because bootstrap block is nil")
```
Any reason to log this at info and not debug like the message directly above does?

tock-ibm replied:
Changed

wlahti commented on integration/e2e/e2e_test.go:
```diff
@@ -489,6 +489,8 @@ var _ = Describe("EndToEnd", func() {
 			ordererRunner := network.OrdererRunner(orderer)
 			ordererProcess = ifrit.Invoke(ordererRunner)
 			Eventually(ordererProcess.Ready, network.EventuallyTimeout).Should(BeClosed())
+			Eventually(ordererRunner.Err(), network.EventuallyTimeout).Should(gbytes.Say("Setting up cluster when there is no system channel: assuming cluster-type consensus"))
```
In the context of each file, these logs seemed to make sense, but it seems a bit redundant to see what's essentially the same information logged three times.

tock-ibm replied:
Alas, that is the way I found to ensure the changes I made to the consenter and main.go are in effect... any suggestions?

wlahti replied:
Perhaps we switch all but one of these messages to debug level and just set the necessary loggers to debug for this test? And maybe include a TODO (or create a JIRA cleanup task) to remove the messages once some of the other functionality is place for us to actually test that the orderer is usable without a system channel.

tock-ibm replied:
I have a solution: I simply verify that the registrar has an etcdraft consenter when it is initialized without a system channel, and throw panic if it does not. This is much more explicit and helps testing the registrar in UT as well.

tock-ibm committed:
Review comments: tune log messages

- Tune log messages
- Check the registrar has an etcdraft consenter when it starts without channels

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I0214949251c14efa68744024b5d655a23aa30ffc

C0rWin commented on orderer/common/server/main.go:
```diff
@@ -113,58 +112,91 @@ func Main() {
+			logger.Info("Bootstrapping the system channel")
+			initializeBootstrapChannel(bootstrapBlock, lf)
+		} else if len(lf.ChannelIDs()) > 0 {
+			logger.Info("Not bootstrapping because of existing channels")
```
I found it very difficult to comprehand this log message, what exactly you are trying to state here?

tock-ibm replied:
I hope this is better:
```go
logger.Infof(
   "Not bootstrapping the system channel because the bootstrap block number is %d (>0), replication is needed", 
   bootstrapBlock.Header.Number)
```

C0rWin commented on orderer/common/server/main.go:
```diff
@@ -113,58 +112,91 @@ func Main() {
 			}
 		}
+	} else {
+		logger.Infof("Setting up cluster without a system channel")
```
Isn't that a duplication of line 138? This also makes me think that in general the flow is pretty tangled here with respect to way we handle different bootstrap procedures. Not sure whenever this PR is the right place, but ideally I think that we should have clearer logic segregation.

tock-ibm replied:
That's true. I refactored the tricky part to be clearer and not duplicate the cluster setup code.

C0rWin commented on orderer/common/server/main.go:
```diff
@@ -113,58 +112,91 @@ func Main() {
+	// This happens only when the orderer is operating with a system channel.
+	if clusterBootBlock != nil {
+		consensusTypeName := consensusType(clusterBootBlock, cryptoProvider)
+		_, clusterType = clusterTypes[consensusTypeName]
```
a very small nit: but IMO `isClusterType` would be better here.

tock-ibm replied:
ok

C0rWin commented on orderer/common/server/main.go:
```diff
@@ -113,58 +112,91 @@ func Main() {
+	} else {
+		logger.Infof("Setting up cluster without a system channel")
+		clusterType = true
+		clusterClientConfig = initializeClusterClientConfig(conf)
```
Since lines 188 thru 199 actually repition of lines 164-175, I think this needs to be extracted into funciton.

tock-ibm replied:
done, removed the duplicate code

C0rWin reviewed (commented):
a few nits.

tock-ibm committed:
Review comments: simplify main

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I237d0cfa018d5689180bcb137ac12d2c5daf3ffc
